Use app.slice by default in user manager (and define special user slices)

benzea opened the pull request:
OK, not sure if we really want this … but it feels sane to me, so I figured I would propose it here.

As the commit says, there are alternatives. In particular as we have currently defined the `app-` prefix for all units, we could move them around based on that (i.e. ship `app-.scope` and `app-.slice` drop-ins). But, that isn't really a good option either for various reasons.

CC: @davidedmundson @smcv @iainlane

smcv commented on man/systemd.special.xml:
```diff
@@ -1167,6 +1167,52 @@
+          <listitem>
+            <para>By default all essential session services and applications
+            are placed into <filename>session.slice</filename>.
+            The exact definition depends on the desktop environment</para>
```
How is that expected to work? For example, dbus-daemon is usually an "essential" part of desktops, so should the maintainer of dbus-daemon be putting `Slice=session.slice` into `dbus.service`? Or is the maintainer of dbus-daemon meant to leave it as-is, and then GNOME, KDE, XFCE etc. are each expected to ship `.d` dropins that move it into `session.slice`?

(Or if you're using dbus-broker, the same questions, but for that.)

benzea replied:
I expected services to opt-in to this. For DBus it should be added to the service file.

Shipping drop-in configurations in each desktop would be quite a pain. And, one installed DE doing this would modify all other ones on the same machine (unless you do ugly shenanigans of creating drop-in configurations at login time). I find it acceptable to do this in the short term, but only if we can expect upstreams to adopt the setting.

I hope that in most cases it should be pretty obvious what should go into `session.slice` (and what is acceptable; e.g. many `gnome-settings-daemon` plugins are not important enough, but we'll probably just place them all into `session.slice` as they are tiny.)

davidedmundson replied:
>I expected services to opt-in to this

That's also how I envisioned things, it's the main reason to upstream the slice definitions into systemd.

smcv replied:
If that's the case, it's probably a good idea to say so, so that user-service authors have a reference for what they should do. Something like:

> Essential session services should be configured to be part of this slice, by adding `Slice=session.slice` to their unit files.

As far as I can see, it's harmless for a service to be assigned to this slice (or to background.slice) in older versions of systemd that don't define it or do anything special to create it, for example if you have a new version of `dbus` running on an old version of `systemd`. Is that true?

davidedmundson replied:
session.slice will be created implicitly 

It just won't have the nice pretty description or documentation link if someone introspects the slice

benzea replied:
I am changing it to (trying to integrate @smcv's suggestion with what was there):

```
            <para>All essential services and applications required for the
            session should use this slice.
            Such services should be configured to be part of this slice by
            adding <varname>Slice=session.slice</varname> to their unit files.</para>
```

smcv replied:
Nice.

davidedmundson commented on man/systemd.special.xml:
```diff
@@ -1167,6 +1167,52 @@
+          <listitem>
+            <para>All services and applications that require running
+            low-priority background tasks should use this slice.
+            This permits allows resources to be preferentially assigned to
```
"This permits allows"  -> "This permits"

smcv commented:
> we have currently defined the app- prefix for all units

Is there a specification for this, perhaps on freedesktop.org somewhere? If there is, a cross-reference to it would be good to have here.

The GNOME Terminal (3.36.x) on my Debian testing/unstable system is currently in:

```
apps.slice -> apps-org.gnome.Terminal.slice -> vte-spawn-${UUID}.scope
apps.slice -> apps-org.gnome.Terminal.slice -> gnome-terminal-server.service
```

Is that an older version of this convention, with `apps` replaced by `app` in newer versions?

smcv commented:
> Is there a specification for this, perhaps on freedesktop.org somewhere?

Ah, I see it's <https://systemd.io/DESKTOP_ENVIRONMENTS/>. I think a cross-reference to that specification would be useful.

davidedmundson commented:
@smcv  Specification is at https://systemd.io/DESKTOP_ENVIRONMENTS/

and yes apps -> app was a change between the various draft specs. (60ca8f22b80255fe10c40017ffe15a9c67b31f1c)

One interesting thing about Ben's proposal of setting a default slice is potentially we could just drop that prefix entirely.

benzea commented:
OK, added a sentence linking to the documentation.

Yep, we changed `apps` to `app`, VTE/gnome-terminal has already been fixed upstream.

Yeah, one motivation for the `app-` prefix was to allow drop-in configurations. We really can consider dropping that requirement if we go down this path. Basically, just mandating that the last (or second to last for scopes) part of the unit name as delimited by `-` should be the application ID.

keszybz commented on man/systemd.special.xml:
```diff
@@ -1167,6 +1167,55 @@
+        <varlistentry>
+          <term><filename>background.slice</filename></term>
+          <listitem>
+            <para>All services and applications that require running
```
All services and applications that require running low-priority background tasks → All services running low-priority background tasks ?

keszybz commented on units/user/systemd-exit.service:
```diff
@@ -14,3 +14,6 @@ DefaultDependencies=no
 SuccessAction=exit-force
+
+[Service]
+Slice=-.slice
```
Hmm, shutdown should be blocked by an .slice unit. Weed to fix that in the manager anyway.

Also:
systemd-tmpfiles-clean.service
systemd-tmpfiles-clean.timer
systemd-tmpfiles-setup.service
→ those should be in the background.slice.

benzea replied:
Only `systemd-tmpfiles-clean.service`, I would say. The timer doesn't have it and `setup` runs early and needs to get its job done. Added a commit to set it.

benzea commented:
> But then this doesn't really work for things which ship their own units, like bus launched applications which specify a unit. How are they supposed to become subject to the desktop's policy?

Yeah, it isn't quite clear to me either …

The idea of using the `app-(launcher-)?` prefix was exactly to permit drop-in configurations. In the case of e.g. flatpak wanting to ship something, it makes things really simple. But if you want to apply a common set of rules across a DE only, then it is messy. It is possible, but to avoid collisions, you need to dynamically install the drop-in for `app-.scope` into `$XDG_RUNTIME_DIR/systemd/user` while the session is running.

Werkov commented on src/core/unit.c:
```diff
@@ -3319,12 +3319,16 @@ int unit_set_default_slice(Unit *u) {
+                                ? SPECIAL_SYSTEM_SLICE
+                                : SPECIAL_ROOT_SLICE;
+                else
+                        slice_name = SPECIAL_APP_SLICE;
```
This default would apply to units that don't specify a slice. In my understanding those would be processes either fitting under `session.slice` (such as dbus daemon, keyring damons/agents) or `background.slice` (custom user-defined jobs, monitoring, e-mail sync). What is the reason for the `app.slice`? (Maybe keeping the `-.slice` would be the best at this moment.)

keszybz replied:
The assumption is that we'll have a limited amount of jobs that go into `session.slice`. The rest of what the user starts fits into either `app` or `background`, and if we pick between the two, `app` is a better choice. This is somewhat similar to how we normally start programs from the command-line: in the foreground by default if nothing else is specified.

Werkov replied:
IIUC, you take in from the POV of the launcher (`StartTransientUnit()` DBus call) -- I agree with the foreground semantics. But what about static (user) unit files? Could it go into NEWS that their default location will be changed?

keszybz replied:
> IIUC, you take in from the POV of the launcher (`StartTransientUnit()` DBus call) -- I agree with the foreground semantics.

I was considering any way to start a service (unit file, systemd-run --user, glib wrappers, whatever). *How* it is started is something of an implementation detail.

> But what about static (user) unit files? Could it go into NEWS that their default location will be changed?

Yep, any change like this will need to be described in NEWS. But we now have the practice that NEWS is constructed before the first prerelease. Trying to update NEWS from individual PRs only leads to an endless series of merge conflicts.

poettering replied:
with the perpetual check in unit_is_extrinsic() (as suggested above) then SPECIAL_INIT_SCOPE will also be considered extrinsic automatically (since it is already perpetual), and the if conditoins will be simpler her.

keszybz commented:
All the bionic-* tests fail the same way:
```
autopkgtest [16:05:14]: test boot-smoke: [-----------------------
reboot #0
bash: line 1: 13286 Killed                  /tmp/autopkgtest.y4Qn8I/build.zGD/systemd/debian/tests/boot-smoke 2> >(tee -a /tmp/autopkgtest.y4Qn8I/boot-smoke-stderr >&2) > >(tee -a /tmp/autopkgtest.y4Qn8I/boot-smoke-stdout)
autopkgtest [16:05:15]: test process requested reboot with marker 1
Exit request sent.
autopkgtest-virt-ssh: WARNING: ssh connection failed. Retrying in 3 seconds...
waiting to boot...
checking for running system
checking for failed unmounts for user systemd
checking for connection timeouts
bash: line 1:   668 Killed                  /tmp/autopkgtest.y4Qn8I/build.zGD/systemd/debian/tests/boot-smoke 2> >(tee -a /tmp/autopkgtest.y4Qn8I/boot-smoke-stderr >&2) > >(tee -a /tmp/autopkgtest.y4Qn8I/boot-smoke-stdout)
checking that NetworkManager runs
322
reboot #1
autopkgtest [16:05:55]: test process requested reboot with marker 2
Exit request sent.
autopkgtest-virt-ssh: WARNING: ssh connection failed. Retrying in 3 seconds...
waiting to boot...
checking for running system
checking for failed unmounts for user systemd
Aug 03 16:06:10 autopkgtest systemd[436]: Failed unmounting /boot/efi.
Aug 03 16:06:10 autopkgtest systemd[436]: Failed unmounting /dev/hugepages.
Aug 03 16:06:10 autopkgtest systemd[436]: Failed unmounting /dev/mqueue.
Aug 03 16:06:10 autopkgtest systemd[436]: Failed unmounting /run/user/1000.
Aug 03 16:06:10 autopkgtest systemd[436]: Failed unmounting /run/user/113.
Aug 03 16:06:10 autopkgtest systemd[436]: Failed unmounting /sys/fs/fuse/connections.
Aug 03 16:06:10 autopkgtest systemd[436]: Failed unmounting /sys/kernel/config.
Aug 03 16:06:10 autopkgtest systemd[436]: Failed unmounting /sys/kernel/debug.
Aug 03 16:06:10 autopkgtest systemd[436]: Failed unmounting /sys/kernel/tracing.
Aug 03 16:06:11 autopkgtest systemd[418]: Failed unmounting /boot/efi.
Aug 03 16:06:11 autopkgtest systemd[418]: Failed unmounting /dev/hugepages.
Aug 03 16:06:11 autopkgtest systemd[418]: Failed unmounting /dev/mqueue.
Aug 03 16:06:11 autopkgtest systemd[418]: Failed unmounting /run/user/113.
Aug 03 16:06:11 autopkgtest systemd[418]: Failed unmounting /sys/fs/fuse/connections.
Aug 03 16:06:11 autopkgtest systemd[418]: Failed unmounting /sys/kernel/config.
Aug 03 16:06:11 autopkgtest systemd[418]: Failed unmounting /sys/kernel/debug.
Aug 03 16:06:11 autopkgtest systemd[418]: Failed unmounting /sys/kernel/tracing.
found failed unmount in journal
autopkgtest [16:06:37]: test boot-smoke: -----------------------]
boot-smoke           FAIL non-zero exit status 1
```

I'll force-push to retrigger the CI.

keszybz commented:
```
--- stderr ---
app.slice: Failed to create cgroup /user.slice/user-1000.slice/session-1.scope/app.slice: Permission denied
app.slice: Failed to create cgroup /user.slice/user-1000.slice/session-1.scope/app.slice: Permission denied
path-exists.service: Succeeded.
app.slice: Failed to create cgroup /user.slice/user-1000.slice/session-1.scope/app.slice: Permission denied
path-exists.service: Succeeded.
path-exists.path: Succeeded.
app.slice: Failed to create cgroup /user.slice/user-1000.slice/session-1.scope/app.slice: Permission denied
app.slice: Failed to create cgroup /user.slice/user-1000.slice/session-1.scope/app.slice: Permission denied
path-existsglob.service: Succeeded.
app.slice: Failed to create cgroup /user.slice/user-1000.slice/session-1.scope/app.slice: Permission denied
path-existsglob.service: Succeeded.
path-existsglob.path: Succeeded.
app.slice: Failed to create cgroup /user.slice/user-1000.slice/session-1.scope/app.slice: Permission denied
app.slice: Failed to create cgroup /user.slice/user-1000.slice/session-1.scope/app.slice: Permission denied
path-changed.service: Succeeded.
app.slice: Failed to create cgroup /user.slice/user-1000.slice/session-1.scope/app.slice: Permission denied
path-changed.service: Succeeded.
path-changed.path: Succeeded.
app.slice: Failed to create cgroup /user.slice/user-1000.slice/session-1.scope/app.slice: Permission denied
app.slice: Failed to create cgroup /user.slice/user-1000.slice/session-1.scope/app.slice: Permission denied
path-modified.service: Succeeded.
app.slice: Failed to create cgroup /user.slice/user-1000.slice/session-1.scope/app.slice: Permission denied
path-modified.service: Succeeded.
path-modified.path: Succeeded.
app.slice: Failed to create cgroup /user.slice/user-1000.slice/session-1.scope/app.slice: Permission denied
app.slice: Failed to create cgroup /user.slice/user-1000.slice/session-1.scope/app.slice: Permission denied
path-mycustomunit.service: Failed to attach to cgroup /user.slice/user-1000.slice/session-1.scope/app.slice/path-mycustomunit.service: No such file or directory
path-mycustomunit.service: Failed at step CGROUP spawning /bin/true: No such file or directory
path-mycustomunit.service: Main process exited, code=exited, status=219/CGROUP
path-mycustomunit.service: Failed with result 'exit-code'.
Test timeout when testing path-unit.path
-------
```

benzea commented:
> app.slice: Failed to create cgroup /user.slice/user-1000.slice/session-1.scope/app.slice: Permission denied

Huh? I am confused, why would anything try to create that cgroup? it should be underneath `user@1000.service` not `session-1.scope`.

keszybz commented:
That test does some low-level cgroup mangling. It's not something that is normally supported, but we do such things in tests... If you don't want too, I can try to look at this later to figure it out.

benzea commented:
> That test does some low-level cgroup mangling. It's not something that is normally supported, but we do such things in tests... If you don't want too, I can try to look at this later to figure it out.

I'll take a look now, I was simply confused and had not even found the appropriate logs pointing to the issue at the time :)

benzea commented:
Hah, that test failure is not new! The test only works most of the times because `Type=simple` means the service is `active` before it reaches the cgroup setup code. If you e.g. `path-directorynotempty.service` change it to `Type=exec` it fails consistently.

So, I see two solutions:
 1. Somehow give the test permissions to manage the cgroup subhierarchy of `session-X.scope` (though setting `Delegate=yes` in a drop-in does not appear to work.
 2. Wrap the test in `systemd-run -d -t --wait --user`, which results in appropriate permissions to be present.

poettering commented on man/systemd.special.xml:
```diff
@@ -1167,6 +1167,53 @@
+        </varlistentry>
+
+        <varlistentry>
+          <term><filename>app.slice</filename></term>
```
bikeshedding: do we really want to abbreviate that? why not `applications.slice`?

benzea replied:
It would be rather painful to change the three slice names at this point :-/

poettering replied:
ok

keszybz replied:
I'd also vote for "app". It's shorter, and "app" is has already become a word of its own. ("App" is maybe not exactly the same as "application", but I don't think this distinction matters in this context.)

poettering commented on man/systemd.special.xml:
```diff
@@ -1167,6 +1167,55 @@
+          <term><filename>session.slice</filename></term>
+          <listitem>
+            <para>All essential services and applications required for the
+            session should use this slice.
```
reading this i am not sure what belongs here and what in app.slice? i.e. one is "essential" the other is not, but what does that mean?

poettering replied:
i.e. maybe add a list of examples here and above. i.e. list "web browser app, text editor app, graphics app" under app.slice and "desktop shell, display server, screen readers, …" or so under "session.slice".

Btw, why not call this "essential.slice"? or "core.slice" or so?

poettering commented on man/systemd.special.xml:
```diff
@@ -1167,6 +1167,53 @@
+          <term><filename>background.slice</filename></term>
+          <listitem>
+            <para>All services running low-priority background tasks should use this slice.
+            This permits resources to be preferentially assigned to the other slices.</para>
```
an example or two for this would be good too. as above. Also maybe say explicitly that the stuff listed here would also fit into session.slice semantically, in most ways except that the stuff should be put at a disadvantaged priority for latency and resource usage

benzea replied:
OK, tried to come up with better strings. I turned it around a bit by focusing a bit on a good definition for`session.slice`.

poettering commented on units/user/background.slice:
```diff
@@ -0,0 +1,12 @@
+#  (at your option) any later version.
+
+[Unit]
+Description=User background tasks slice
```
so far we uppercased the first letter of the words in our Description= strings (applies to all three units, of course)

poettering commented on src/core/unit.c:
```diff
@@ -3316,12 +3316,16 @@ int unit_set_default_slice(Unit *u) {
+                        slice_name =
+                                !unit_has_name(u, SPECIAL_INIT_SCOPE)
+                                ? SPECIAL_SYSTEM_SLICE
+                                : SPECIAL_ROOT_SLICE;
```
inverted conditions are weird. when you touch this maybe swap the order of the ternary op here, i.e. 

```c
                        slice_name =
                                unit_has_name(u, SPECIAL_INIT_SCOPE)
                                ? SPECIAL_ROOT_SLICE
                                : SPECIAL_SYSTEM_SLICE;
```

poettering commented on units/user/systemd-exit.service:
```diff
@@ -14,3 +14,6 @@ DefaultDependencies=no
 SuccessAction=exit-force
+
+[Service]
+Slice=-.slice
```
I think this fix is actually right, but could you add a comment why you added this to the unit file

poettering reviewed (changes requested):
approach looks good to me, some comments

keszybz commented:
The CI should pass on the next push now.

poettering commented:
I wonder if we should have "background.slice" also for system service stuff.

poettering commented:
pr lgtm to me too

Werkov commented:
> I wonder if we should have "background.slice" also for system service stuff.
I'd leave that to distros or even local admins. (But it's apparently OT
in this thread.)

keszybz commented:
Pfff, various autopkgtests are failing in boot-smoke. Unfortunately that test is completely inscrutable. I'd be inclined to ignore it, unless somebody finds an actual fault in this PR.

/cc @xnox

xnox commented:
> Pfff, various autopkgtests are failing in boot-smoke. Unfortunately that test is completely inscrutable. I'd be inclined to ignore it, unless somebody finds an actual fault in this PR.
> 
> /cc @xnox

@rbalint I wonder if we should be moving testing up from bionic to focal now. And also, I guess trying to figure out what's going here?

rbalint commented:
> > Pfff, various autopkgtests are failing in boot-smoke. Unfortunately that test is completely inscrutable. I'd be inclined to ignore it, unless somebody finds an actual fault in this PR.
> > /cc @xnox
> 
> @rbalint I wonder if we should be moving testing up from bionic to focal now. And also, I guess trying to figure out what's going here?

Yes, absolutely. Master does not target Bionic, for sure. OTOH i386 is an interesting question ...

keszybz commented:
I pushed two new commits that will hopefully fix the test failure (the immediate cause of the failure is fixed, but I didn't actually run the tests). PTAL.

Werkov commented:
I find the extrinsic distinction a step in a good direction. As you wrote Slice= for such units isn't well defined, so be it `-.slice` with the comment.
Ad ordering issues -- do the extrinsic units need to be included in any (user instance) transactions at all?

keszybz commented:
>  Ad ordering issues -- do the extrinsic units need to be included in any (user instance) transactions at all?

The units are available to create dependencies on external state: for example, user units can be started when devices are plugged in, user units can be ordered after mounts. At least that's how I understand it. I think it doesn't work in practice, because the user manager tries to mount/unmount things on its own and fails with permission errors. I even wrote a patch to reject those attempts, but then I realized that there are mount units which may be managed from the user instance, for example when `user` is used in /etc/fstab or when udisks are used to elevate privileges. So I think we need to do some cleanup here, but that's a separate issue for this PR.

poettering commented on src/core/swap.c:
```diff
@@ -56,6 +56,38 @@ static bool SWAP_STATE_WITH_PROCESS(SwapState state) {
+        if (MANAGER_IS_SYSTEM(u->manager))
+                return false;
+
+        return true;
```
```c
return MANAGER_IS_USER(u->manager);
```

keszybz replied:
I styled this on the surrounding functions, which do the same pattern. But this version works too.

poettering commented on src/core/mount.c:
```diff
@@ -422,10 +423,10 @@ static bool mount_is_extrinsic(Mount *m) {
                 return true;
 
-        if (UNIT(m)->perpetual) /* All perpetual units never change state */
+        if (u->perpetual) /* All perpetual units never change state */
```
let's move this check into unit_is_extrinsic(), it's is generically true.

poettering commented on src/core/unit.h:
```diff
@@ -684,6 +687,10 @@ int unit_set_description(Unit *u, const char *description);
 bool unit_may_gc(Unit *u);
 
+static inline bool unit_is_extrinsic(Unit *u) {
+        return UNIT_VTABLE(u)->is_extrinsic && UNIT_VTABLE(u)->is_extrinsic(u);
```
as suggested above, make this:

```
return u->perpetual ||
        (UNIT_VTABLE(u)->is_extrinsic && UNIT_VTABLE(u)->is_extrinsic(u));
```

poettering reviewed (commented):
looks good, some comments

(we should expose the perpetual + extrinsic properties on the bus, one day)

keszybz commented:
Updated as requested.

poettering commented on src/core/unit.c:
```diff
@@ -3350,15 +3350,19 @@ int unit_set_default_slice(Unit *u) {
+        else if (MANAGER_IS_SYSTEM(u->manager)) {
+                if (unit_has_name(u, SPECIAL_INIT_SCOPE))
+                        slice_name = SPECIAL_ROOT_SLICE;
                 else
```
you can drop these three lines now.

keszybz replied:
Indeed. Updated.
